File: iOS/Kalorie/Core/UseCases/FoodItemSubmissionFetcher.swift
//
//  FoodItemSubmissionFetcher.swift
//  Kalorie
//
//  Created by Josef Antoni on 11.09.2026.
//

import Foundation

enum FoodItemSubmissionFetcher {

    // MARK: - Functions

    static func fetch(
        where field: String,
        isEqualTo value: String,
        dataProvider: any FirestoreDataProviderProtocol
    ) async throws -> [FoodItemSubmissionDomain] {
        let dtos: [FoodItemSubmissionDTO] = try await dataProvider.loadAsync(
            from: Constants.Firestore.foodItemSubmissions,
            where: field,
            isEqualTo: value,
            orderBy: "submitted_at",
            descending: true
        )
        return dtos.map { $0.asDomain() }
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/usecases/FoodItemSubmissionFetcher.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.models.FoodItemSubmissionDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.FoodItemSubmissionDTO
import antoni.kalorie.core.networking.loadAsync
import antoni.kalorie.core.utils.Constants

object FoodItemSubmissionFetcher {

    // MARK: - Functions

    suspend fun fetch(field: String, isEqualTo: String, dataProvider: FirestoreDataProviderProtocol): List<FoodItemSubmissionDomain> {
        val dtos: List<FoodItemSubmissionDTO> = dataProvider.loadAsync(
            from = Constants.Firestore.FOOD_ITEM_SUBMISSIONS,
            field = field,
            isEqualTo = isEqualTo,
            orderBy = "submitted_at",
            descending = true,
        )
        return dtos.map { it.asDomain() }
    }
}
